writer.go

```diff
@@ -0,0 +1,317 @@
+)
+
+// WriterConfig is a struct containing the configurable parameters for a Writer
+type WriterConfig struct {
+	// BatchDocumentCount is the max document count to send in one request. Must be less than or equal to MaxDocumentCount
+	BatchDocumentCount uint64
+
+	// TODO implement document size restriction.
+
+	// Workers number of workers to use, defaults to 1.
+	Workers uint64
+	// FlushInterval is the longest interval to wait before flushing buffered documents.
+	FlushInterval time.Duration
+}
+
+func (c WriterConfig) Validate() error {
+	if c.BatchDocumentCount < 1 {
+		return fmt.Errorf("document count must be greater than 0")
+	}
+	if c.BatchDocumentCount > MaxDocumentCount {
+		return fmt.Errorf("document count must be less than or equal to %d", MaxDocumentCount)
+	}
+
+	if c.FlushInterval == time.Duration(0) {
+		return fmt.Errorf("batch size must be greater than 0")
+	}
+
+	return nil
+}
+
+// Writer is a helper that writes structs to Rockset collections
+type Writer struct {
+	adder  DocumentAdder
+	config WriterConfig
+	stats  WriteStats
+
+	buffers      map[string]map[string][]interface{}
+	bufferedDocs uint64
+
+	m       sync.Mutex
+	timeout *time.Timer
+
+	wg      sync.WaitGroup
+	workers int
+
+	addDocRequests chan addDocRequest
+	writeRequests  chan WriteRequest
+	stop           chan struct{}
+}
+
+// WriteRequest contains the data to be written to a Rockset collection
+type WriteRequest struct {
+	Workspace  string
+	Collection string
+	Data       interface{}
+}
+
+type addDocRequest struct {
+	workspace  string
+	collection string
+	data       []interface{}
+}
+
+// WriteStats holds counters for the documents written to Rockset
+type WriteStats struct {
+	DocumentCount uint64
+	ErrorCount    uint64
+}
+
+// DocumentAdder is the interface used to write documents to Rockset, and is implemented by the RockClient.
+type DocumentAdder interface {
+	AddDocuments(ctx context.Context, workspace, collection string, docs []interface{}) ([]openapi.DocumentStatus, error)
+}
+
+// NewWriter creates a new Writer
+func NewWriter(conf WriterConfig, client DocumentAdder) (*Writer, error) {
+	// set defaults if the zero value is used
+	if conf.Workers == 0 {
+		conf.Workers = 1
+	}
+	if conf.BatchDocumentCount == 0 {
+		conf.BatchDocumentCount = DefaultDocumentCount
+	}
+	if conf.FlushInterval == 0 {
+		conf.FlushInterval = DefaultFlushInterval
+	}
+
+	if err := conf.Validate(); err != nil {
+		return nil, err
+	}
+	return &Writer{
+		adder:          client,
+		wg:             sync.WaitGroup{},
+		writeRequests:  make(chan WriteRequest, conf.BatchDocumentCount),
+		addDocRequests: make(chan addDocRequest, conf.Workers),
+		stop:           make(chan struct{}, 1),
+		buffers:        make(map[string]map[string][]interface{}),
+		config:         conf,
+		stats:          WriteStats{},
+	}, nil
+}
+
+// C returns the WriteRequest channel, which is used to send documents to be added to Rockset.
+func (w *Writer) C() chan<- WriteRequest {
```

Comment:
Small nit, should this have a better name?

Reply:
Possibly, but I don't know what it should be, and `C` is some sort of convention, e.g. the [time.Ticker](  https://pkg.go.dev/time#Ticker) struct in the standard library
```
type Ticker struct {
	C <-chan [Time](https://pkg.go.dev/time#Time) // The channel on which the ticks are delivered.
	// contains filtered or unexported fields
}
```